Contacts menu

ChristophWurst opened the pull request:
Rough todo list:
- client-side
  - [x] display actions listed by the API
  - [x] ~~lazy loading~~ not necessary if number of elements is limited to 25.
  - [x] link to contacts app
  - [x] search
    - [x] show input
    - [x] query and show results from server
- server-side
  - [x] ~~discuss API design in separate issue~~ got not feedback within two weeks
  - [x] implement API (registration system + some example clients, e.g. in Mail) Example: https://github.com/nextcloud/contacts/pull/173

implements https://github.com/nextcloud/server/issues/207

<hr>

Current screenshot:
![bildschirmfoto von 2017-03-06 15-46-58](https://cloud.githubusercontent.com/assets/1374172/23614701/2e4289d2-0284-11e7-9dd7-8012795b5fea.png)

ChristophWurst committed:
Contacts menu

* load list of contacts from the server
* show last message of each contact

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
Dang it. I've lost my most recent changes due to a rebase+force push of an outdated version of the branch 😠

ChristophWurst commented:
> Dang it. I've lost my most recent changes due to a rebase+force push of an outdated version of the branch 😠

Got them back from my other PC 💪 😌

nickvergessen commented on lib/private/Contacts/ContactsMenu/ActionProviderStore.php:
```diff
@@ -0,0 +1,84 @@
+					'message' => "Could not load contacts menu action provider $class",
+					'app' => 'core',
+				]);
+				throw new \Exception("Could not load contacts menu action provider $class");
```
Don't throw `$class` to avoid input injection. You log the class, that's fine enough

ChristophWurst replied:
Makes sense!

LukasReschke commented on core/templates/layout.user.php:
```diff
@@ -163,5 +170,14 @@
 			</div>
 		</div>
+
+		<script nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>" type="text/javascript">
```
Doing this here has some security implications as a XSS in `$_['content']` may be able to extract this by using an `<img>` tag etc.

Can we move that to a dedicated script of before the content? :)

ChristophWurst replied:
Sure we can. I just wanted to do it inline for performance reasons as we'd save one request per page load.

LukasReschke replied:
IMHO better solved by making sure we concatenate all JS files that we deliver anyways on nearly every page load :)

ChristophWurst commented:
```
1) Tests\Contacts\ContactsMenu\EntryTest::testAddAndSortAction
usort(): Array was modified by the user comparison function
/drone/src/github.com/nextcloud/server/lib/private/Contacts/ContactsMenu/Entry.php:105
/drone/src/github.com/nextcloud/server/lib/private/Contacts/ContactsMenu/Entry.php:87
/drone/src/github.com/nextcloud/server/tests/lib/Contacts/ContactsMenu/EntryTest.php:112
```
php5.6+usort+phpunit mocks = 💩

ChristophWurst commented:
Alright, this has already gotten bigger than I wanted it to be for the first PR. Let's do the more advanced stuff in follow-up PRs:
* Avatar from the contacts app
* App action provider plugin system
* Last interaction

ChristophWurst commented:
@nextcloud/designers I need some help with styling the different elements:
![bildschirmfoto von 2017-02-20 22-04-21](https://cloud.githubusercontent.com/assets/1374172/23141769/b497bcc2-f7b8-11e6-936c-3a73031fdcd4.png)

Should the inner popover menu overlap the scrollbar or should it only be visible inside the menu?

@skjnldsv do you know if I have to/should tweak the css to get the popover menu positioned right or are there any server classes I have to use or guidelines I have to follow? :-)

skjnldsv commented:
@ChristophWurst https://github.com/nextcloud/server/issues/2798

skjnldsv commented:
Just want to say that it looks damn good! 🦀

skjnldsv commented:
By the way the other menu on the header don't use a popovermenu.
The contacts one looks a bit out of place compared to the others :)

skjnldsv commented:
> try to comply with the css guidelines

😂 
I will try to help on this after my work :)

jancborchardt commented:
Heya, answering some of your questions @ChristophWurst 

> Should the inner popover menu overlap the scrollbar or should it only be visible inside the menu?

- [ ] It’s fine to overlap the scrollbar, it’s a level over the inner menu. It should also be above the scrollable area because when you open the menu on an entry down at the end it should show properly and not expand inside the scrollable area.

> @skjnldsv: By the way the other menu on the header don't use a popovermenu.

- [ ] Yup, we need to adjust the detail positioning here. No space between header and menu, and remove border-radius on the top corners. :)

- [ ] The icon needs to be also available in white, depending on theme. cc @juliushaertl 

- [ ] The max-height of the popover should be exactly n+1/2 of an entry, so you see when there’s more. Maybe 4 1/2 is best?

- [ ] On scrolling down, new contacts should be loaded automatically via lazy loading. For the future when you have chats with all these folks I think it’s fine to just have all of it in there without a shortlink to Contacts. You will be able to use search anyway.


Anything else? I guess lots is for a follow-up PR anyway. :)

MorrisJobke commented:
One notices that you work in Firefox. All the weird rendering behaviour that only Firefox does are used by your CSS 😉 

* [ ] SVG is too big: 
<img width="152" alt="bildschirmfoto 2017-02-21 um 18 02 54" src="https://cloud.githubusercontent.com/assets/245432/23190901/6b0fc83a-f860-11e6-86c6-1777630f52a9.png">
* [ ] menu is somehow too small
<img width="217" alt="bildschirmfoto 2017-02-21 um 18 03 04" src="https://cloud.githubusercontent.com/assets/245432/23190906/7636ec7a-f860-11e6-819d-702ec8b539c6.png">

This works fine in Firefox, but not in Chrome nor Safari.

MorrisJobke commented:
I fixed the icon color and it now looks a lot nicer:

<img width="453" alt="bildschirmfoto 2017-02-21 um 18 15 33" src="https://cloud.githubusercontent.com/assets/245432/23191186/e2027432-f861-11e6-9883-9227b2281d82.png">

MariusBluem commented:
@MorrisJobke Better, but should the item indicate about whether it is opened or not (a bit more grey or something...) 🤔 @jancborchardt

skjnldsv commented:
Disagree @MariusBluem . We have the triangle indicator, that's enough I think :)

LukasReschke commented on lib/private/Contacts/ContactsMenu/Manager.php:
```diff
@@ -0,0 +1,101 @@
+		return [
+			'contacts' => $topEntries,
+			'contactsAppEnabled' => $contactsEnabled,
+			'contactsAppURL' => $contactsURL,
```
Will generate a URL without index.php for me which breaks. You should either link to the route directly, or always prepend `/index.php/` or just use the JS `OC.generateUrl('/apps/contacts');` which should take care of that.

In this scenario I'd probably vote for the JS approach.

LukasReschke commented on lib/private/Contacts/ContactsMenu/Providers/DetailsProvider.php:
```diff
@@ -0,0 +1,71 @@
+
+		// TODO: unique contact URL to the contacts app
+		// TODO: l10n
+		$contactsUrl = $this->urlGenerator->getAbsoluteURL('/apps/contacts');
```
Same as https://github.com/nextcloud/server/pull/3233/files#r102418886

skjnldsv commented:
I merged the other properties we use in the user menu, apps menu, notifications menu into a single (already existing) class: 'menu'. This was the one who handled the arrow too.
Now this fit right.

- [ ] Close the menu when clicking outside.

skjnldsv commented:
@ChristophWurst do you register the menu like the guidelines (which aren't written anywhere 😂) seems to be? `OC.registerMenu($('#contactsmenu-trigger'), $('#contactsmenu-menu'));`
https://github.com/nextcloud/server/blob/354bbd50a78b49d349c92b2bca584b9585369d05/core/js/js.js#L1353

ChristophWurst commented:
> @ChristophWurst do you register the menu like the guidelines (which aren't written anywhere 😂) seems to be?

Thanks for the tip. This works nicely 🚀

ChristophWurst commented on core/js/contactsmenu.js:
```diff
@@ -0,0 +1,473 @@
+		+ '<a class="top-action {{contact.topAction.icon}}" href="{{contact.topAction.hyperlink}}"></a>'
+		+ '{{#if contact.actions.length}}'
+		+ '    <span class="other-actions icon-more"></span>'
+		+ '    <div class="menu">'
```
Nested menus don't seem to work. Do I have to adjust the CSS for this or should it work out of the box? 😉 
![bildschirmfoto von 2017-03-06 10-09-27](https://cloud.githubusercontent.com/assets/1374172/23603342/15578726-0255-11e7-99f3-dcd538dad697.png)

ChristophWurst commented:
review time!

jancborchardt commented:
Fixed some small CSS issues, here’s some interaction things which remain:

- [ ] The triangle isn’t placed correctly for long names, needs to be done via JS as we do in the apps menu
- [ ] The message »Loading your contacts …« appears every single time for 1–2 seconds and makes it appear slow
- [ ] When in the Mail app, the height of the initial loading spinner makes the popup much higher than elsewhere.
- [ ] Contacts are sorted alphabetically. We don’t have anything like favoriting or another mechanism, do we? @Henni @irgendwie @skjnldsv 
- [ ] Can we at least check who you have a call with in the Video calls app and sort those up top?
- [ ] When only 2 icons are there, we can show both (like Mail & Details). Because otherwise we have the situation where there’s a mail icon, and then a 3-dot menu just for »Details«.
- [ ] When clicking on »Details«, only the main Contacts app view is shown, not the contact details
- [ ] When clicking on »Mail«, and you have the Mail app installed and an account, that should be used. :)
- [ ] Can we add a video calls icon? cc @Ivansss @nickvergessen. At the very basis this would be creating a room and then sending them an email with the link as an easy flow.
- [ ] Searching is a bit strange.
  - [ ] feels a tad slow
  - [ ] there is a big spinner replacing the content while searching – that spinner should only be in the search field on the right
  - [ ] when the search is carried out (like when you typed a few characters) the cursor is set to the beginning of the input. That messes with your search term ;)
  - [ ] »No contacts found« should have an emptycontent icon, ideally icon-search.

Sorry for the long list :\ really awesome work, and I’ll try to fix some more of the CSS issues. :)

Also cc @nextcloud/designers @nextcloud/javascript for review and help! :rocket:

ChristophWurst commented:
> The message »Loading your contacts …« appears every single time for 1–2 seconds and makes it appear slow

We so have to fetch the data from the server.

> When in the Mail app, the height of the initial loading spinner makes the popup much higher than elsewhere.

Maybe a css rule of the Mail app overwrites the spinner position? Should be easy to check ;-)

> Can we at least check who you have a call with in the Video calls app and sort those up top?

We can of course improve this feature in further iterations. I wouldn't add such an enhanced sorting mechanism now, it's already pretty big and hard to review.

> When clicking on »Details«, only the main Contacts app view is shown, not the contact details

Known "bug" -> documented in the code. Basically the contacts app will later provider the entry for the menu (code moved over) because it also knows the correct URL to the specific contact. I don't want to introduce a dependency on contacts.

> When clicking on »Mail«, and you have the Mail app installed and an account, that should be used. :)

It is, if you've configured the mailto handler correctly. Of course, Mail can later hook into this and overwrite the default browser behavior. This is however, out of scope for this PR and the server component.

> Can we add a video calls icon? cc @Ivansss @nickvergessen. At the very basis this would be creating a room and then sending them an email with the link as an easy flow.

Yes, but also just inside the app. See my comments above, the app registration mechanism can be added later. All apps will get their chance to inject whatever crazy interaction they want to enable you to do with your contacts.

> feels a tad slow

We have to query the server. There's very little we can do about that. Of course, client-side search could be performed in the meantime, but this can be tricky and I'd rather do that as a further enhancement and not in this initial version.

nickvergessen commented:
> > Can we add a video calls icon? cc Ivansss nickvergessen. At the very basis this would be creating a room and then sending them an email with the link as an easy flow.
> 
> Yes, but also just inside the app. See my comments above, the app registration mechanism can be added later. All apps will get their chance to inject whatever crazy interaction they want to enable you to do with your contacts.

I was about to say, when there is an API point that allows us to do stuff we surely can.

jancborchardt commented:
> > The message »Loading your contacts …« appears every single time for 1–2 seconds and makes it appear slow
> 
> We so have to fetch the data from the server.

Can we do that on page load already?


> 
> > When in the Mail app, the height of the initial loading spinner makes the popup much higher than elsewhere.
> 
> Maybe a css rule of the Mail app overwrites the spinner position? Should be easy to check ;-)

Yep, will look into this. :)

> > Can we at least check who you have a call with in the Video calls app and sort those up top?
> 
> We can of course improve this feature in further iterations. I wouldn't add such an enhanced sorting mechanism now, it's already pretty big and hard to review.

Ok! :)


> > When clicking on »Details«, only the main Contacts app view is shown, not the contact details
> 
> Known "bug" -> documented in the code. Basically the contacts app will later provider the entry for the menu (code moved over) because it also knows the correct URL to the specific contact. I don't want to introduce a dependency on contacts.

Does »later« mean still for Nextcloud 12? Same as for Video calls it would be strange if one of the main apps doesn’t work correctly with this. :) cc @Henni @irgendwie @skjnldsv 


> > When clicking on »Mail«, and you have the Mail app installed and an account, that should be used. :)
> 
> It is, if you've configured the mailto handler correctly. Of course, Mail can later hook into this and overwrite the default browser behavior. This is however, out of scope for this PR and the server component.

Right, makes sense.

> > Can we add a video calls icon? cc @Ivansss @nickvergessen. At the very basis this would be creating a room and then sending them an email with the link as an easy flow.
> 
> Yes, but also just inside the app. See my comments above, the app registration mechanism can be added later. All apps will get their chance to inject whatever crazy interaction they want to enable you to do with your contacts.

Ok, but by »later« do you mean still in Nextcloud 12? It would be a bit strange to have our own main communication app not supported. ;) cc @nickvergessen @Ivansss 


> > feels a tad slow
> 
> We have to query the server. There's very little we can do about that. Of course, client-side search could be performed in the meantime, but this can be tricky and I'd rather do that as a further enhancement and not in this initial version.

Ok – I think the other points are more important for perceived performance anyway. :)

ChristophWurst commented:
> Fixed some small CSS issues, here’s some interaction things which remain:

Well, apparently you "fixed" it only for Chome:
![bildschirmfoto von 2017-03-13 08-31-04](https://cloud.githubusercontent.com/assets/1374172/23845286/7f3bffc4-07c7-11e7-9bff-771a12987fa4.png)
FF:
![bildschirmfoto von 2017-03-13 08-30-59](https://cloud.githubusercontent.com/assets/1374172/23845287/7f41eb64-07c7-11e7-9543-357fb091a923.png)

😉

ChristophWurst commented:
>>> The message »Loading your contacts …« appears every single time for 1–2 seconds and makes it appear slow
>>>
>> We so have to fetch the data from the server.
>>
> Can we do that on page load already?

Sure we can. But then the client would be out of sync with the server state, like newly added contacts are missing, updated attributes are not synced, last interaction (which will be added later) won't be up to date. Our technology stack doesn't allow to do real-time like applications easily. Of course we could add some kind of polling in the background, but this is - again - tricky and I wouldn't add it in the first iteration.

In my testing setups it loads pretty fast btw. Any idea where you could have a bottleneck? Maybe too many contacts? I have a few hundreds here, this doesn't seem to be a problem so far.

ChristophWurst commented:
@jancborchardt I took care of the reported issues where I thought they are needed for the first working version. Let me know if anything's missing to get this in :-)

@nickvergessen any advice on how I could let apps specify the action's icon? Now they pass in the icon class, e.g. 'icon-search'. I'm not sure if this is the most flexible way. I thought about passing the absolute URL of the icon, but then none of the existing CSS rules are applied of course :-/

nickvergessen commented:
> any advice on how I could let apps specify the action's icon? Now they pass in the icon class, e.g. 'icon-search'. I'm not sure if this is the most flexible way. I thought about passing the absolute URL of the icon, but then none of the existing CSS rules are applied of course :-/

Yeah I would recommend urls instead of CSS, because otherwise all apps will load their CSS file on each page etc. Look at the activity filters, or admin sections for examples.

Speaking of them, I would also recommend to register those plugins via info.xml 🙊

ChristophWurst commented:
> Speaking of them, I would also recommend to register those plugins via info.xml

Sure, that's the plan (also noted somewhere in the many comments above 😉)

If everybody is fine with reviewing 5k lines of dif I can add this in this PR. Otherwise I'd prefer to add the app registration mechanism in a follow-up PR.

jancborchardt commented:
@ChristophWurst here’s the remarks I have left before merging:

- [ ] The menu doesn’t actually show the avatars, only the placeholders.
- [ ] Scrolling down and then simply clicking into the search box brings up »Loading contacts« and scrolls you back to the top. It shouldn’t do anything
- [ ] In the screenshot you showed above the Mail icon is shown in the row and then again in the menu. It shouldn’t be duplicated. It doesn’t happen for me anyway, but just making sure. :)
- [ ] For me now with contacts who have no mail address, this is how the menu looks like: 
![capture du 2017-03-15 20-16-18](https://cloud.githubusercontent.com/assets/925062/23966861/304cb7d0-09bd-11e7-920e-a79ef2645ce6.png)
- [ ] The triangle position needs to be set via Javascript (just like for the app menu) in the case when you have no avatar set and thus your name is shown: 
![capture du 2017-03-15 20-20-24](https://cloud.githubusercontent.com/assets/925062/23966886/48b9445a-09bd-11e7-8a7b-e1982e865cab.png)
- [ ] And again of course the Details link should be fixed before a Nextcloud 12 release, but as far as I understand this needs to be done in the Contacts app?

I’m going to do some more CSS fixes. :)

ChristophWurst commented:
> In the screenshot you showed above the Mail icon is shown in the row and then again in the menu. It shouldn’t be duplicated. It doesn’t happen for me anyway, but just making sure. :)

Well, how many mail addresses did that contact have? I guess just one 😉

ChristophWurst commented:
> The layout of »Loading your contacts« and »Looking for [contact]« should be the same as the emptycontent: First the loading spinner on top, then the text below. Then it looks less twitchy when it changes. :) (Let me know where I can set a breakpoint or so so I can test or change the code)

Please take a look. Some styles interfere there, hence it looks a bit odd. Just add a `return;` statement before https://github.com/nextcloud/server/pull/3233/files#diff-bb881a25a51a93d9dee391bc5a92316cR488 and it should "load forever". Thanks for helping with this!

ChristophWurst commented:
> Scrolling down and then simply clicking into the search box brings up »Loading contacts« and scrolls you back to the top. It shouldn’t do anything

The problem here is the *clear input* button that only appears on Chrome. We want it to work, but it doesn't send any event when it clears the input, hence I have to listen for a click event. Dunno how to solve this therefore. If I remove the click handler the menu doesn't reflect the cleared input state.

ChristophWurst commented:
> Then maybe in the menu the text should be not »Mail« but the actual email address. And the primary one should show the email address which is behind it via tooltip.js?

I'm not entirely happy with this approach, since tooltips do not work on mobile devices.

jancborchardt commented:
@ChristophWurst if you rebase then I can fix the style. :) And then we could merge also so we can really test with everything. :)

jancborchardt commented:
@rullzer just as a reminder – something broke after the rebase – please fix. :) I fixed the placement of the arrow.

rullzer commented:
@jancborchardt fixed. Work your magic!

ChristophWurst committed:
Show mail address in popover menu

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst committed:
Use tooltop for contact's top action title

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
@jancborchardt please take a look at the loading/no content layout differences and the newly added tooltip for the top action. Thanks 😉

ChristophWurst committed:
Let apps register contact menu provider via info.xml

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
Added app provider registration. One can now test this in conjunction with https://github.com/nextcloud/contacts/pull/173 🚀

ChristophWurst commented:
Wrote a little app that shows how to integrate other actions (links to Twitter profile): https://github.com/ChristophWurst/contactsmenu_social

jancborchardt commented:
@ChristophWurst the list shouldn’t show yourself, nor the »i« Details icon for internal users. :) (At least not yet where we don’t have a detail page for them.)

ChristophWurst commented:
> @ChristophWurst the list shouldn’t show yourself, nor the »i« Details icon for internal users. :) (At least not yet where we don’t have a detail page for them.)

Did you actually check out the branch? 😛 Only contacts of the contacts app should have the details action. I already addressed that in nextcloud/contacts#173 😉

jancborchardt commented:
@ChristophWurst of course, correct branch on both server and Contacts. :) Still occurs. I can see myself in the list (including detail action) and also local users _with_ the detail action – leading to the Contacts app.

ChristophWurst commented:
> @ChristophWurst of course, correct branch on both server and Contacts. :) Still occurs. I can see myself in the list (including detail action) and also local users with the detail action – leading to the Contacts app.

Congrats, than you've found the first bug in nextcloud/contacts#173 😉

ChristophWurst committed:
Don't show contacts an entry for themselves

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst committed:
Use absolute URI for action icons

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

MorrisJobke commented:
* [ ] the contact menu looks a bit off:

<img width="389" alt="bildschirmfoto 2017-04-11 um 12 28 11" src="https://cloud.githubusercontent.com/assets/245432/24922031/76529416-1eb2-11e7-86c4-61cd548bc2cb.png">

The white area is a link that looks like this:

```html
<a class="second-action http://192.168.99.100/server/core/img/actions/info.svg" href="http://192.168.99.100/server/index.php/apps/contacts"></a>
```

I guess the link in the `class` attribute is wrong there 😉

ChristophWurst committed:
Fix icon of second action

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
@MorrisJobke fixed :)

MorrisJobke reviewed (approved):
Looks good 👍

MorrisJobke commented:
> @MorrisJobke fixed :)

Conflicts ;)

ChristophWurst commented:
> Conflicts ;)

rebased :)

rullzer commented:
So if I don't have the mail app, or contacts/calendar. Then what does the menu do? Besides showing me the contacts...

ChristophWurst commented:
It lets you email people. Mail app is not required for that.

rullzer commented:
Ok so what I did:

1. Create users `admin`, `user1`, `user2`
2. Click on the contacts menu

It just shows me:
![foo](https://cloud.githubusercontent.com/assets/45821/25284615/d89d4d2a-26b7-11e7-9c90-04cc7a7083bb.png)

With no possibility to click

MorrisJobke commented:
> With no possibility to click

Do you contacts have an email address set? 🙈 It  only showed up in those cases for me 😉

rullzer commented:
@MorrisJobke nope. Would not make sense otherwise. But my point is that if you have contacts that have no means of communication set this menu looks silly

MorrisJobke commented:
> @MorrisJobke nope. Would not make sense otherwise. But my point is that if you have contacts that have no means of communication set this menu looks silly

 would keep it that way for now anyways, because more actions should be added in the future.

MorrisJobke commented:
> would keep it that way for now anyways, because more actions should be added in the future.

See #4374 for example

rullzer commented:
ok. well then fine by me. Lets merge and fix tiny bugs later!

rullzer commented:
but... failing tests

LukasReschke commented on core/routes.php:
```diff
@@ -57,6 +57,7 @@
 		['name' => 'Preview#getPreview', 'url' => '/core/preview.png', 'verb' => 'GET'],
 		['name' => 'Css#getCss', 'url' => '/css/{appName}/{fileName}', 'verb' => 'GET'],
 		['name' => 'Js#getJs', 'url' => '/js/{appName}/{fileName}', 'verb' => 'GET'],
+		['name' => 'contactsMenu#index', 'url' => '/contactsmenu/contacts', 'verb' => 'GET'],
```
POST? 😄

ChristophWurst replied:
Why?

LukasReschke replied:
Not awesome to have such stuff in the access log 😉

ChristophWurst replied:
Sure I can change that, although I don't really get why. This route queries data from the server, but does not change any state. Hence I find it uncommon to use POST here. Please elaborate.

jancborchardt commented:
Can we just agree to not do a last-minute merge before feature freeze? ;) Let’s postpone further nitpicking after it’s merged so it’s not piled up even more.

MorrisJobke commented:
> Can we just agree to not do a last-minute merge before feature freeze? ;) Let’s postpone further nitpicking after it’s merged so it’s not piled up even more.

Okay, this then means no contacts menu in 12. Fine by me.

cc @karlitschek

jancborchardt commented:
Not sure why this means that. Apparently it's just a failing test and Lukas' remark about GET vs POST?

MorrisJobke commented:
> Not sure why this means that. Apparently it's just a failing test and Lukas' remark about GET vs POST?

You just said "no last minute merges before the feature freeze". This means we will not merge it for the 12 feature freeze and then have it in 13. Or did I understand this wrong?

MorrisJobke commented:
And if you mean with "no last-minute merges" following "don't merge before tests are passing and all remarks are addressed": Of course we will not merge something that breaks tests.

karlitschek commented:
We definitely have to merge this for 12. I agree that smaller improvements can be done later. Even after the feature freeze.

ChristophWurst committed:
Fix failing contacts menu js test data

When we test wheter action menus in the contacts menu close
when clicking other ones, we have to provide test data
that actually causes the view to render the menu.

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst committed:
Adjust entry unit test to newly added avatar property

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
Both js and php tests should pass now, but let's see what Drone reports.

ChristophWurst committed:
Update composer autoloader

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst committed:
Use HTTP POST to retrieve the list of contacts

We do not want to have sensitive information in the URL and
therefore also not in the access log. Thus the GET request is
replaced by a POST request.

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

nickvergessen commented on core/routes.php:
```diff
@@ -57,6 +57,7 @@
 		['name' => 'Preview#getPreview', 'url' => '/core/preview.png', 'verb' => 'GET'],
 		['name' => 'Css#getCss', 'url' => '/css/{appName}/{fileName}', 'verb' => 'GET'],
 		['name' => 'Js#getJs', 'url' => '/js/{appName}/{fileName}', 'verb' => 'GET'],
+		['name' => 'contactsMenu#index', 'url' => '/contactsmenu/contacts', 'verb' => 'POST'],
```
👎 filters/searches always use GET (e.g. share sidebar, etc.), should also be the case here.

ChristophWurst replied:
See https://github.com/nextcloud/server/pull/3233#pullrequestreview-34154411 and https://github.com/nextcloud/server/pull/3233/commits/c852349edda5b4abfe9a6d9d40a700f1890da8b4. This is @LukasReschke's call.

nickvergessen replied:
Ah , didn't see the comment, I saw the commit and that caught me ;)

ChristophWurst commented:
@LukasReschke @MorrisJobke @nickvergessen any other remarks?

jancborchardt reviewed (approved):
Works nicely now! :+1:

MorrisJobke commented:
Loading panel looks a bit broken in Safari:

![ohne titel](https://cloud.githubusercontent.com/assets/245432/25368028/382bfa1c-2950-11e7-944b-98af37629187.png)


Only looks slightly better in Chrome (but still looks weird):

![ohne titel 2](https://cloud.githubusercontent.com/assets/245432/25368035/4288d372-2950-11e7-9f31-a52016d9279d.png)


@jancborchardt Could you recheck this one?

jancborchardt committed:
Fix spinner position of Contacts menu

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt committed:
Fix emptycontent HTML and wording

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt commented:
Fixed and adjusted the loading emptycontent style to the »No contacts found« layout so it looks less janky.

cc @MorrisJobke @ChristophWurst :)

jancborchardt commented:
@ChristophWurst @MorrisJobke is that Drone failure related or not?

@LukasReschke @nickvergessen any other remarks?

ChristophWurst commented:
> @ChristophWurst @MorrisJobke is that Drone failure related or not?

Probably unrelated, but I've rebased onto latest master to have the fix for the failing js unit tests. Now everything *should* pass.

jospoortvliet commented:
@ChristophWurst there's a conflict with the header.scss now, though... I tried to fix it but the menu just disappeared.

ChristophWurst commented:
Rebased. Thanks for testing, @jospoortvliet! :)

jospoortvliet commented:
Tested your rebased version, works too.
![spectacle w25913](https://cloud.githubusercontent.com/assets/551757/25402887/89899004-29fa-11e7-868d-d95f4bf24ac1.png)
The 'more apps' menu is totally broken, though ;-)
![spectacle k25913](https://cloud.githubusercontent.com/assets/551757/25402901/9546e144-29fa-11e7-891c-e779f729aebf.png)
@jancborchardt probably YOUR FAULT 😬

jancborchardt committed:
Fix test

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt committed:
Merge branch 'master' into contactsmenu

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt commented:
So I
- fixed the test (which was failing due to a text change I made)
- fixed merge conflicts in PHP files

I think it goes without saying this should be double-checked. :joy:

beerlao commented:
Wouldn't it make sense to at least give an option to hide local users in contacts menu? Then only contacts from contacts app would be shown. Otherwise users will be able to see contact details and contact images of other users on the same NC12 instance.

ChristophWurst commented:
@ChristophWurst please document the API in the dev manual

ChristophWurst commented:
> @ChristophWurst please document the API in the dev manual

https://github.com/nextcloud/documentation/pull/11222
